### ISSUE TITLE
Find a macOS embedding path that beats the CPU

### ISSUE THREAD

**Issue by dmwyatt** [dmwyatt/granz#142]
## Problem

macOS release builds shipped `--features coreml`, but CoreML never beat the CPU on
nomic-embed-text-v1.5 — it lost to it, badly. That feature is being removed from the
macOS release build in a companion PR, which makes plain CPU the fastest thing grans
can do on Apple Silicon. This issue tracks finding something better.

## Evidence

Measured on an M1 Max against the exact model grans caches
(`fastembed_cache/models--nomic-ai--nomic-embed-text-v1.5/.../onnx/model.onnx`),
batch 8 x seq 512, via onnxruntime directly:

| Provider | Run time | Throughput |
|---|---|---|
| CPU | 2,516 ms | 3.18 chunks/s |
| CoreML NeuralNetwork (what we shipped) | 41,041 ms | 0.19 chunks/s |
| CoreML MLProgram | 8,594 ms | 0.93 chunks/s |
| CoreML MLProgram + CPUAndGPU | 7,850 ms | 1.02 chunks/s |
| CoreML MLProgram + CPUAndANE | 3,821 ms | 2.09 chunks/s |

Every CoreML configuration loses. The shipped one (NeuralNetwork, the ort default) was
between 3x and 16x slower than CPU across repeated runs — it is high-variance, but it
never won. At batch 1 the gap narrows and CoreML still does not win.

## Why CoreML loses

The EP's own capability report on this graph:

```
number of partitions supported by CoreML: 73
number of nodes in the graph: 1192
number of nodes supported by CoreML: 470
```

39% of nodes supported, split across **73 partitions**. Every partition boundary is a
hand-off between CoreML and the CPU EP with tensor copies, so a single inference becomes
73+ round trips and the fragmentation overhead swamps any compute win.

The specific blockers, from verbose EP logs:

- Nomic's rotary embeddings emit zero-sized dimensions, which CoreML rejects outright:
  `CoreML does not support shapes with dimension values of 0. Input:/encoder/layers.0/attn/rotary_emb/Slice_5_output_0, shape: {-1,-1,12,0}`
- Core transformer ops are unsupported under dynamic shapes:
  `Operator [MatMul] is not supported by the impl`, likewise `Cast`, `Reshape`, `Gather`,
  `Slice`, `Unsqueeze`, `Concat`

This is a model/EP mismatch, not a misconfiguration. CoreML's NeuralNetwork format wants
static shapes; a dynamic-shape transformer with rotary embeddings is near a worst case.

## Avenues worth trying

Roughly in order of effort-to-payoff:

1. **Quantized model on CPU.** fastembed exposes `NomicEmbedTextV15Q`. Cheapest thing to
   measure, no new dependencies, and int8 on CPU may well beat anything CoreML does.
   Requires a quality check: re-embedding under a different model changes vectors, so it
   needs a `benchmark quality` run against the frozen snapshot before adoption.
2. **A static-shape ONNX export.** Fix the sequence length (512) and re-export so the
   rotary-embedding zero-dims and dynamic-shape MatMuls disappear. If the partition count
   collapses from 73 to single digits, CoreML becomes worth re-measuring. Costs padding
   waste on short chunks, which may eat the win.
3. **An export without the rotary pattern**, or a different embedding model whose graph
   is CoreML-friendly. Largest change: a model swap re-opens the Phase 0-5 quality work
   in `docs/search-roadmap.md`.
4. **Truncate to 256d.** Not an acceleration path for the forward pass, but the roadmap
   already flags nomic v1.5's Matryoshka training as the first lever for vector-side cost.

## Gate

Anything adopted here needs a real before/after, not a microbenchmark: end-to-end
`grans embed` throughput on a real corpus, plus `benchmark quality` against the frozen
snapshot if the vectors change at all. Options 1 and 3 change vectors; option 2 does not.

## Related

- #7 (detect and report which execution provider is active) — this is the case #7 was
  worried about, inverted. The provider bound successfully and reported nothing wrong; it
  was simply slower. Provider reporting alone would not have caught this, so whatever #7
  ships should surface throughput, not just provider names.

**Comment by dmwyatt** [comment]
## Measured on an M4: nothing under ONNX Runtime beats the CPU

Rented a Scaleway M4-SP Mac mini (Apple M4, 10C CPU / 10C GPU / 16-core ANE, 16 GB, macOS 26.6.1) for a day and ran `grans embed` end to end, as the gate asks, rather than a microbenchmark. Corpus is synthetic (200 generated meetings, 33,789 utterances, 6,470 chunks; a 25-meeting / 796-chunk subset for screening), so throughput is real and nothing sensitive left the machine. Builds are release builds of `3623cc0` with small experiment-only patches: an env var to pick the fastembed model, and an env var to set CoreML EP options (`ModelFormat::MLProgram`, `ComputeUnits`, `with_static_input_shapes`, model cache).

### Full corpus, nomic-embed-text-v1.5, 6,470 chunks

| Run | Seconds | Chunks/s | vs CPU |
|---|---|---|---|
| CPU fp32, batch 16 | 369 | 17.5 | baseline |
| CPU fp32, batch 8 / 32 | 360 / 376 | 18.0 / 17.2 | same |
| int8 `NomicEmbedTextV15Q`, CPU, batch 16 | 362 | 17.9 | same |
| CoreML default (NeuralNetwork, dynamic shapes) | 1,098 | 5.9 | 3.0x slower |
| CoreML MLProgram + CPUAndNeuralEngine | 1,069 | 6.1 | 2.9x slower |

### Screening subset, 796 chunks, CoreML variants for nomic

| Run | Seconds |
|---|---|
| CPU fp32 | 47 |
| CoreML MLProgram (dynamic) | 579 |
| CoreML MLProgram + CPUAndGPU | 100 |
| CoreML NeuralNetwork + static input shapes | 48 |
| CoreML MLProgram + ANE + static input shapes, batch 1 / 4 / 16 / 64 | 44 / 43 / 47 / 48 |

Static input shapes were the one thing with a chance, since they target the 73-partition fragmentation directly. They do remove the penalty, but only to parity. Sampling process CPU utilization mid-run shows why: the "CoreML static" runs burn 550-780% CPU, the same as the pure CPU runs (550-705%), and take the same time. With static shapes the CoreML partitions stop costing anything, but the CPU execution provider is still doing the work; the ANE is not.

### Other fastembed models, 796 chunks (option 3 screen, speed only)

| Model | CPU | CoreML default | CoreML static + ANE |
|---|---|---|---|
| nomic-embed-text-v1.5 (768d) | 47 s | ~140 s (3x) | 47 s |
| BGE-base-en-v1.5 (768d) | 36 s | 752 s (21x) | 36 s |
| BGE-small-en-v1.5 (384d) | 16 s | 67 s (4x) | 16 s |
| all-MiniLM-L6-v2 (384d) | 9 s | 34 s (4x) | 8 s |

Same shape for every model, rotary embeddings or not: dynamic-shape CoreML is a large loss, static-shape CoreML is parity with the CPU and never faster. The rotary pattern makes nomic's case worse but is not the root cause; ORT's CoreML bridge does not accelerate any of these BERT-family graphs.

### Conclusions against the four avenues

1. Quantized model on CPU: no gain on Apple Silicon (362 s vs 369 s). Not worth the vector change.
2. Static-shape export: neutralizes CoreML rather than making it win. Parity with CPU, so no reason to ship it.
3. Different model: does not unlock the ANE/GPU either. The only speed available from a swap is "smaller model is faster on CPU" (BGE-base 1.3x, BGE-small 3x, MiniLM 5x), which is a quality question, not an acceleration path. Quality benchmarking of the fast models is being run separately on the frozen snapshot.
4. 256d truncation: unaffected by any of this; remains a vector-side lever only.

Batch size is irrelevant on CPU (8/16/32 within noise).

The practical answer: under ONNX Runtime, CPU is the ceiling on Apple Silicon, at ~17.5 chunks/s on an M4 (expect roughly half that on an M1). Getting this model onto the GPU/ANE would need a native backend (coremltools conversion or PyTorch MPS), a second embedding runtime for a one-time re-embed, which is not worth it for this tool. The regression that motivated this issue (shipping CoreML enabled, 3x slower) is already fixed by #143. Closing.

Related: #7 should report throughput, not just provider names; every CoreML configuration here bound successfully and reported nothing wrong.

**Comment by dmwyatt** [comment]
## Follow-up: quality cost of the faster models

Since a model swap is the only thing that buys speed on Apple Silicon (smaller model, same CPU), I ran the three screened models through `benchmark quality` on the frozen snapshot (872 docs, 93-query v2 golden set, k=10), each embedded into a snapshot copy with an experiment-only env-var patch. Nomic baseline re-run the same day and matches the ledger exactly.

Semantic mode (retriever alone), hit-rate / recall / MRR @10:

| Model | CPU speed vs nomic (M4) | hit | recall | MRR |
|---|---|---|---|---|
| nomic-embed-text-v1.5 | 1.0x | 86.0% | 75.9% | 0.715 |
| BGE-base-en-v1.5 | 1.3x | 82.8% | 70.2% | 0.634 |
| BGE-small-en-v1.5 | 3x | 84.9% | 62.2% | 0.591 |
| all-MiniLM-L6-v2 | 5x | 84.9% | 69.4% | 0.711 |

End to end with the shipping pipeline (hybrid + rerank-jina + title boost):

| Model | hit | recall | MRR | vs nomic (recall / MRR) |
|---|---|---|---|---|
| nomic-embed-text-v1.5 | 93.5% | 80.5% | 0.813 | baseline |
| BGE-base-en-v1.5 | 93.5% | 77.6% | 0.795 | -2.9 / -0.018 |
| BGE-small-en-v1.5 | 92.5% | 74.5% | 0.763 | -6.0 / -0.050 |
| all-MiniLM-L6-v2 | 88.2% | 74.8% | 0.768 | -5.7 / -0.045 |

The reranker recovers some MRR but not recall: what the retriever never surfaces in its top pool cannot be reranked back. MiniLM is the interesting one (semantic MRR at parity with nomic at 5x the speed) but it loses 5 to 6 points of recall and 5 points of hit-rate end to end, and paraphrase queries suffer most. None of the three is an acceptable trade for a one-time re-embed cost; nomic stays.

Ledger entries recorded with note `model swap experiment: <model>, snapshot copy, 2026-08-28`.

**Comment by dmwyatt** [comment]
Follow-up: the ONNX/CoreML framing was the wrong question. llama.cpp on Metal runs the same nomic model at ~17k tok/s on an M1 Max, about 3x the best ONNX CPU number above. Continued in #144.
